demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file mqtt_subscription_manager.c
+ * @brief Implementation of the API of a subscription manager for handling subscription callbacks
+ * to topic filters in MQTT operations.
+ */
+
+/* Standard includes. */
+#include <string.h>
+#include <assert.h>
+
+/* Include header for the subscription manager. */
+#include "mqtt_subscription_manager.h"
+
+/**
+ * @brief Represents a registered record of the topic filter and its associated callback
+ * in the subscription manager registry.
+ */
+typedef struct SubscriptionManager_Record
+{
+    const char * pTopicFilter;
+    uint16_t topicFilterLength;
+    SubscriptionManager_Callback_t callback;
+} SubscriptionManager_Record_t;
+
+/**
+ * @brief The default value for the maximum size of the callback registry in the
+ * subscription manager.
+ */
+#ifndef MAX_SUBSCRIPTION_CALLBACK_RECORDS
+    #define MAX_SUBSCRIPTION_CALLBACK_RECORDS    5
+#endif
+
+/**
+ * @brief The registry to store records of topic filters and their subscription callbacks.
+ */
+static SubscriptionManager_Record_t callbackRecordList[ MAX_SUBSCRIPTION_CALLBACK_RECORDS ] = { 0 };
+
+/**
+ * @brief Handle special corner cases regarding wildcards at the end of topic
```

Comment:
The description is vague. Please elaborate the description.

Reply:
Updated description to mention the special cases handled in the utility